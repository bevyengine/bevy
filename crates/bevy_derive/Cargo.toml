[package]
name = "bevy_derive"
version = "0.2.1"
edition = "2018"
authors = [
    "Bevy Contributors <bevyengine@gmail.com>",
    "Carter Anderson <mcanders1@gmail.com>",
]
description = "Provides derive implementations for Bevy Engine"
homepage = "https://bevyengine.org"
repository = "https://github.com/bevyengine/bevy"
license = "MIT"
keywords = ["bevy"]

[lib]
proc-macro = true

[dependencies]
Inflector = { version = "0.11.4", default-features = false }
proc-macro-crate = "0.1.5"
proc-macro2 = "1.0"
quote = "1.0"
syn = "1.0"
