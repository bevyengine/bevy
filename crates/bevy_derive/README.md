# Bevy Derive

[![License](https://img.shields.io/badge/license-MIT%2FApache-blue.svg)](https://github.com/bevyengine/bevy#license)
[![Crates.io](https://img.shields.io/crates/v/bevy_derive.svg)](https://crates.io/crates/bevy_derive)
[![Downloads](https://img.shields.io/crates/d/bevy_derive.svg)](https://crates.io/crates/bevy_derive)
[![Docs](https://docs.rs/bevy_derive/badge.svg)](https://docs.rs/bevy_derive/latest/bevy_derive/)
[![Discord](https://img.shields.io/discord/691052431525675048.svg?label=&logo=discord&logoColor=ffffff&color=7389D8&labelColor=6A7EC2)](https://discord.gg/bevy)
