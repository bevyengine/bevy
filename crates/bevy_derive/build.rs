fn main() {
    println!(
        "cargo:rustc-env=PROC_ARTIFACT_DIR={}",
        std::env::var("OUT_DIR").unwrap()
    )
}
