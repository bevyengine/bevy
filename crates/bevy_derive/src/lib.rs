extern crate proc_macro;

mod app_plugin;
mod as_vertex_buffer_descriptor;
mod bytes;
mod modules;
mod render_resource;
mod render_resources;
mod resource;
mod shader_defs;
mod type_uuid;

use proc_macro::TokenStream;

/// Derives the FromResources trait. Each field must also implement the FromResources trait or this will fail. FromResources is
/// automatically implemented for types that implement Default.
#[proc_macro_derive(FromResources, attributes(as_crate))]
pub fn derive_from_resources(input: TokenStream) -> TokenStream {
    resource::derive_from_resources(input)
}

/// Derives the Bytes trait. Each field must also implements Bytes or this will fail.
#[proc_macro_derive(Bytes, attributes(as_crate))]
pub fn derive_bytes(input: TokenStream) -> TokenStream {
    bytes::derive_bytes(input)
}

/// Derives the RenderResources trait. Each field must implement RenderResource or this will fail.
/// You can ignore fields using `#[render_resources(ignore)]`.
#[proc_macro_derive(RenderResources, attributes(render_resources, as_crate))]
pub fn derive_render_resources(input: TokenStream) -> TokenStream {
    render_resources::derive_render_resources(input)
}

/// Derives the RenderResource trait. The type must also implement `Bytes` or this will fail.
#[proc_macro_derive(RenderResource, attributes(as_crate))]
pub fn derive_render_resource(input: TokenStream) -> TokenStream {
    render_resource::derive_render_resource(input)
}

/// Derives the ShaderDefs trait. Each field must implement ShaderDef or this will fail.
/// You can ignore fields using `#[shader_defs(ignore)]`.
#[proc_macro_derive(ShaderDefs, attributes(shader_def, as_crate))]
pub fn derive_shader_defs(input: TokenStream) -> TokenStream {
    shader_defs::derive_shader_defs(input)
}

/// Derives the AsVertexBufferDescriptor trait.
#[proc_macro_derive(AsVertexBufferDescriptor, attributes(vertex, as_crate))]
pub fn derive_as_vertex_buffer_descriptor(input: TokenStream) -> TokenStream {
    as_vertex_buffer_descriptor::derive_as_vertex_buffer_descriptor(input)
}

/// Generates a dynamic plugin entry point function for the given `Plugin` type.  
#[proc_macro_derive(DynamicPlugin)]
pub fn derive_dynamic_plugin(input: TokenStream) -> TokenStream {
    app_plugin::derive_dynamic_plugin(input)
}

// From https://github.com/randomPoison/type-uuid
#[proc_macro_derive(TypeUuid, attributes(uuid))]
pub fn type_uuid_derive(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    type_uuid::type_uuid_derive(input)
}

#[proc_macro]
pub fn external_type_uuid(tokens: proc_macro::TokenStream) -> proc_macro::TokenStream {
    type_uuid::external_type_uuid(tokens)
}
