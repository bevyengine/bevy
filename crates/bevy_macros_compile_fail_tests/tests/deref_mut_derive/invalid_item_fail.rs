use bevy_derive::DerefMut;

#[derive(DerefMut)]
//~^ ERROR: cannot be derived on field-less structs
struct UnitStruct;

#[derive(DerefMut)]
//~^ ERROR: can only be derived on structs
enum Enum {}
