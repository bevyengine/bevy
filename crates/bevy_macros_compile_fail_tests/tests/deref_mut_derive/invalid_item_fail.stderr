error: DerefMut cannot be derived on field-less structs
 --> tests/deref_mut_derive/invalid_item_fail.rs:3:10
  |
3 | #[derive(DerefMut)]
  |          ^^^^^^^^
  |
  = note: this error originates in the derive macro `DerefMut` (in Nightly builds, run with -Z macro-backtrace for more info)

error: DerefMut can only be derived on structs
 --> tests/deref_mut_derive/invalid_item_fail.rs:7:10
  |
7 | #[derive(DerefMut)]
  |          ^^^^^^^^
  |
  = note: this error originates in the derive macro `DerefMut` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors

